app/eth2wrap/synthproposer.go

```diff
@@ -0,0 +1,205 @@
+// Copyright © 2022 Obol Labs Inc.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package eth2wrap
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/binary"
+	"fmt"
+	"sync"
+
+	eth2client "github.com/attestantio/go-eth2-client"
+	eth2v1 "github.com/attestantio/go-eth2-client/api/v1"
+	eth2p0 "github.com/attestantio/go-eth2-client/spec/phase0"
+	shuffle "github.com/protolambda/eth2-shuffle"
+)
+
+// fifoEpochMax limits the amount of epochs to cache.
+const fifoEpochMax = 10
+
+type synthProposerEth2Provider interface {
+	eth2client.ValidatorsProvider
+	eth2client.SlotsPerEpochProvider
+	eth2client.ProposerDutiesProvider
+}
+
+// synthProposerCache returns a new cache for synthetic proposer duties.
+func newSynthProposerCache(pubkeys []eth2p0.BLSPubKey) *synthProposerCache {
+	return &synthProposerCache{
+		pubkeys:     pubkeys,
+		duties:      make(map[eth2p0.Epoch][]*eth2v1.ProposerDuty),
+		synths:      make(map[eth2p0.Epoch]map[eth2p0.Slot]bool),
+		shuffleFunc: eth2Shuffle,
+	}
+}
+
+// synthProposerCache caches actual and synthetic proposer duties for the set of public keys.
+//
+// Since only a single validator can be a proposer per slot, we require all
+// validators to calculate the synthetic duties for the whole set.
+type synthProposerCache struct {
+	pubkeys []eth2p0.BLSPubKey
+	// shuffleFunc deterministically shuffles the validator indices for the epoch.
+	shuffleFunc func(eth2p0.Epoch, []eth2p0.ValidatorIndex) []eth2p0.ValidatorIndex
+
+	mu     sync.RWMutex
+	fifo   []eth2p0.Epoch
```

Comment:
nit: rename `fifo` to simply `queue`

Reply:
it isn't a queue, it is a reference who was first in and who should be first out.